flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/FileChannelMemoryMappedBoundedData.java

```diff
@@ -0,0 +1,190 @@
+		final long bytesWritten = BufferReaderWriterUtil.writeToByteChannelIfBelowSize(
+				fileChannel, buffer, headerAndBufferArray, spaceLeft);
+
+		if (bytesWritten >= 0) {
+			pos += bytesWritten;
+			return true;
+		}
+		else {
+			return false;
+		}
+	}
+
+	@Override
+	public BoundedData.Reader createReader() {
+		checkState(!fileChannel.isOpen());
+
+		final List<ByteBuffer> buffers = memoryMappedRegions.stream()
+				.map((bb) -> bb.duplicate().order(ByteOrder.nativeOrder()))
+				.collect(Collectors.toList());
+
+		return new MemoryMappedBoundedData.BufferSlicer(buffers);
+	}
+
+	/**
+	 * Finishes the current region and prevents further writes.
+	 * After calling this method, further calls to {@link #writeBuffer(Buffer)} will fail.
+	 */
+	@Override
+	public void finishWrite() throws IOException {
+		mapRegionAndStartNext();
+		fileChannel.close();
+	}
+
+	/**
+	 * Closes the file and unmaps all memory mapped regions
+	 * After calling this method, access to any ByteBuffer obtained from this instance
+	 * will cause a segmentation fault.
+	 */
+	public void close() throws IOException {
+		IOUtils.closeQuietly(fileChannel);
+
+		for (ByteBuffer bb : memoryMappedRegions) {
+			PlatformDependent.freeDirectBuffer(bb);
+		}
+		memoryMappedRegions.clear();
+
+		// To make this compatible with all versions of Windows, we must wait with
+		// deleting the file until it is unmapped.
+		// See also https://stackoverflow.com/questions/11099295/file-flag-delete-on-close-and-memory-mapped-files/51649618#51649618
+
+		Files.delete(filePath);
+	}
+
+	@Override
+	public long getSize() {
+		return pos;
+	}
+
+	private void mapRegionAndStartNext() throws IOException {
+		final ByteBuffer region = fileChannel.map(MapMode.READ_ONLY, startOfCurrentRegion, pos - startOfCurrentRegion);
+		region.order(ByteOrder.nativeOrder());
+		memoryMappedRegions.add(region);
+
+		startOfCurrentRegion = pos;
+		endOfCurrentRegion = startOfCurrentRegion + maxRegionSize;
+	}
+
+	private void throwTooLargeBuffer(Buffer buffer) throws IOException {
+		throw new IOException(String.format(
+				"The buffer (%d bytes) is larger than the maximum size of a memory buffer (%d bytes)",
+				buffer.getSize(), maxRegionSize));
+	}
+
+	// ------------------------------------------------------------------------
+	//  Factories
+	// ------------------------------------------------------------------------
+
+	/**
+	 * Creates new MemoryMappedBuffers, creating a memory mapped file at the given path.
+	 */
+	public static FileChannelMemoryMappedBoundedData create(Path memMappedFilePath) throws IOException {
+		return createWithRegionSize(memMappedFilePath, Integer.MAX_VALUE);
+	}
+
+	/**
+	 * Creates new MemoryMappedBuffers, creating a memory mapped file at the given path.
+	 * Each mapped region (= ByteBuffer) will be of the given size.
+	 */
+	public static FileChannelMemoryMappedBoundedData createWithRegionSize(Path memMappedFilePath, int regionSize) throws IOException {
```

Comment:
`checkNotNull(memMappedFilePath)`, considering this factory might be used in other places, not only for current `Integer.MAX_VALUE` as default size. Is it worth also checking the region size?